### PR TITLE
Fix: io timeout when snell reuse connection

### PR DESCRIPTION
**Pull request by kamingchan**
This bug can be reproduced using [open-snell](https://github.com/icpz/open-snell) and [hey](https://github.com/rakyll/hey).

*hey is a tiny program that sends some load to a web application.*

---

snell server listens at port 5678.

```ini
[snell-server]
listen = 0.0.0.0:5678
psk = psk
obfs = http
```

clash listens at port 9999 with a default outbound to snell server.

```yaml
port: 9999
proxies:
  - name: PROXY
    obfs-opts:
      host: docs.aws.amazon.com
      mode: http
    psk: psk
    server: 127.0.0.1
    port: 5678
    type: snell
    version: 2
rules:
  - MATCH,PROXY
```

use hey to send some http request via clash

```bash
./hey -disable-keepalive -x http://127.0.0.1:9999 https://www.qq.com
```

```
Status code distribution:
  [200]	49 responses

Error distribution:
  [124]	Get "https://www.qq.com": EOF
  [1]	Get "https://www.qq.com": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51282->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51294->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51311->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51336->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51343->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51371->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51381->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51396->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51401->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51424->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51435->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51437->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51441->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51447->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51462->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51476->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51484->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51494->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51505->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51506->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51516->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51537->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51554->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51571->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51622->127.0.0.1:9999: read: connection reset by peer
  [1]	Get "https://www.qq.com": read tcp 127.0.0.1:51623->127.0.0.1:9999: read: connection reset by peer
```

### CODE REVIEW

File: component/snell/pool.go [hunk 1]
```diff
@@ -61,6 +62,7 @@ func (pc *PoolConn) Write(b []byte) (int, error) {
 }
 
 func (pc *PoolConn) Close() error {
+	pc.Snell.Conn.SetReadDeadline(time.Time{})
```
Comment (Dreamacro) [inline component/snell/pool.go:65]: Has this been verified to be fixed? Is it `SetReadDeadline` or `SetDeadline`? Should follow pool timeout? https://github.com/Dreamacro/clash/blob/02842b4c923b90943cd28a9fcced03f84759e4e5/component/snell/pool.go#L75

Comment (kamingchan) [reply component/snell/pool.go:65]: https://github.com/Dreamacro/clash/blob/4e9e4b6cde5d40066d7112add42b5943af4608e5/tunnel/connection.go#L168

clash use `SetReadDeadline` to break the other goroutine when copying bidirectionally between client connection and server connection. Further `Reader.Read()` will return an io timeout error unless read deadline is reseted.

Comment (Dreamacro) [reply component/snell/pool.go:65]: So why isn't `SetReadDeadline(time.Now().Add(15 * time.Second))`

Comment (kamingchan) [reply component/snell/pool.go:65]: https://github.com/Dreamacro/clash/blob/baf03b81e36e3c5d8b9a370ebb82dd2b25b4a58d/component/pool/pool.go#L57

https://github.com/Dreamacro/clash/blob/baf03b81e36e3c5d8b9a370ebb82dd2b25b4a58d/component/snell/pool.go#L76

Any idle tcp connection for more than 15s will be closed by pool. It seems that setting read deadline here is meaningless. Besides if we set a read deadline, we need to reset it when trying to get a connection from pool.

Comment (Dreamacro) [reply component/snell/pool.go:65]: fine, Is it possible to add a line of comment to explain the behavior here?

**Comment by Dreamacro** [comment]
LGTM, thanks!

**Pull request merged**